src/MagnumPlugins/KtxImporter/KtxImporter.cpp

```diff
@@ -0,0 +1,553 @@
+        return;
+    }
+    const auto levelIndex = Containers::arrayCast<Implementation::KtxLevel>(f->in.slice(offset, offset + levelIndexSize));
+    for(Implementation::KtxLevel& level: levelIndex) {
+        Utility::Endianness::littleEndianInPlace(level.byteOffset, level.byteLength, level.uncompressedByteLength);
+        CORRADE_INTERNAL_ASSERT(header.supercompressionScheme != 0 || level.byteLength == level.uncompressedByteLength);
+        CORRADE_INTERNAL_ASSERT(level.uncompressedByteLength % (numFaces * numLayers) == 0);
+        if(data.size() < level.byteOffset + level.byteLength) {
+            Error{} << "Trade::KtxImporter::openData(): mip data too short, expected at least" << level.byteOffset + level.byteLength << "bytes but got" << data.size();
+            return;
+        }
+    }
+
+    f->imageData = Containers::Array<Containers::Array<File::LevelData>>{numLayers};
+
+    for(UnsignedInt layer = 0; layer != numLayers; ++layer) {
+        f->imageData[layer] = Containers::Array<File::LevelData>{numFaces * numMipmaps};
+        /* This matches the image order of DdsImporter: faces, mipmaps */
+        std::size_t currentLevel = 0;
+        for(UnsignedInt face = 0; face != numFaces; ++face) {
+            Vector3i mipSize{size};
+
+            /* Load all mipmaps for current face */
+            for(const Implementation::KtxLevel& level: levelIndex) {
+                const std::size_t length = f->compressed
+                    ? imageLength(mipSize, f->pixelFormat.compressed)
+                    : imageLength(mipSize, f->pixelFormat.uncompressed);
+
+                /* Each mipmap byte range contains tightly packed images,
+                   ordered as follows:
+                   layers, faces, (slices, rows, columns) */
+                const std::size_t imageOffset = (layer * numFaces + face) * length;
+                if(level.byteLength < imageOffset + length) {
+                    Error{} << "Trade::KtxImporter::openData(): image data too short, expected at least" << imageOffset + length << "bytes but got" << level.byteLength;
+                    return;
+                }
+
+                f->imageData[layer][currentLevel++] = {mipSize, f->in.suffix(level.byteOffset).slice(imageOffset, imageOffset + length)};
+
+                /* Shrink to next power of 2 */
+                mipSize = Math::max(mipSize >> 1, Vector3i{1});
+            }
+        }
+    }
+
+    /** @todo Read KTXorientation and flip image? What is the default? */
+    /** @todo Read KTXswizzle and apply it. Can't do that for block-compressed formats, just ignore in that case? */
+    /** @todo Read KTXanimData and expose frame time between images through
+              doImporterState (same as StbImageImporter). What if we need
+              doImporterState for something else, like other API format
+              enum values (KTXdxgiFormat__ etc.)? */
```

Comment:
Huh, KTX supports animations, even? :O Cool.

The animation "support" in StbImageImporter was an ugly quick hack for the [Animated GIF example](https://doc.magnum.graphics/magnum/examples-animated-gif.html) that shouldn't even get made this way.

I think what could work is adding `Trade::AnimationTrackTargetType::ImageLayer` (better name?) and expose the keyframes via `animation()`... and do the same fix for `StbImageImporter` (although they [plan to drop support for animated GIFs](https://twitter.com/nothings/status/1411264318100807681) as it's too much of a nightmare). Though, this is all quite tangential and personally I wouldn't even bother with this unless you have ane asy way to generate the test data for this and want to use them somewhere :)

Reply:
I won't insist on this being added 😄 Just wrote down all my thoughts while reading the spec.

Reply:
Okay, let's ignore animations for now, then :) For fun I might try porting StbImageImporter to use the animation APIs and then that could eventually get mirrored here.